vault/identity_store_entities.go

```diff
@@ -0,0 +1,463 @@
+		},
+		{
+			Pattern: "entity/merge/id/?$",
+			Fields: map[string]*framework.FieldSchema{
+				"from_entity_ids": {
+					Type:        framework.TypeCommaStringSlice,
+					Description: "Entity IDs which needs to get merged",
+				},
+				"to_entity_id": {
+					Type:        framework.TypeString,
+					Description: "Entity ID into which all the other entities needs to get merged",
+				},
+			},
+			Callbacks: map[logical.Operation]framework.OperationFunc{
+				logical.UpdateOperation: i.pathEntityMergeID,
+			},
+
+			HelpSynopsis:    strings.TrimSpace(entityHelp["entity-merge-id"][0]),
+			HelpDescription: strings.TrimSpace(entityHelp["entity-merge-id"][1]),
+		},
+	}
+}
+
+// pathEntityMergeID merges two or more entities into a single entity
+func (i *identityStore) pathEntityMergeID(req *logical.Request, d *framework.FieldData) (*logical.Response, error) {
+	toEntityID := d.Get("to_entity_id").(string)
+	if toEntityID == "" {
+		return logical.ErrorResponse("missing entity id to merge to"), nil
+	}
+
+	fromEntityIDs := d.Get("from_entity_ids").([]string)
+	if len(fromEntityIDs) == 0 {
+		return logical.ErrorResponse("missing entity ids to merge from"), nil
+	}
+
+	toEntityForLocking, err := i.memDBEntityByID(toEntityID)
+	if err != nil {
+		return nil, err
+	}
+
+	if toEntityForLocking == nil {
+		return logical.ErrorResponse("entity id to merge to is invalid"), nil
+	}
+
+	// Acquire the lock to modify the entity storage entry to merge to
+	toEntityLock := locksutil.LockForKey(i.entityLocks, toEntityForLocking.ID)
+	toEntityLock.Lock()
+	defer toEntityLock.Unlock()
+
+	// Create a MemDB transaction to merge entities
+	txn := i.db.Txn(true)
+	defer txn.Abort()
+
+	// Re-read post lock acquisition
+	toEntity, err := i.memDBEntityByID(toEntityID)
+	if err != nil {
+		return nil, err
+	}
+
+	if toEntity == nil {
+		return logical.ErrorResponse("entity id to merge to is invalid"), nil
+	}
+
+	if toEntity.ID != toEntityForLocking.ID {
+		return logical.ErrorResponse("acquired lock for an undesired entity"), nil
+	}
+
+	for _, fromEntityID := range fromEntityIDs {
+		if fromEntityID == toEntityID {
+			return logical.ErrorResponse("to_entity_id should not be present in from_entity_ids"), nil
+		}
+
+		lockFromEntity, err := i.memDBEntityByID(fromEntityID)
+		if err != nil {
+			return nil, err
+		}
+
+		if lockFromEntity == nil {
+			return logical.ErrorResponse("entity id to merge from is invalid"), nil
+		}
+
+		// Acquire the lock to modify the entity storage entry to merge from
+		fromEntityLock := locksutil.LockForKey(i.entityLocks, lockFromEntity.ID)
+
+		// There are only 256 lock buckets and the chances of entity ID collision
+		// is fairly high. When we are merging entities belonging to the same
+		// bucket, multiple attempts to acquire the same lock should be avoided.
+		if fromEntityLock != toEntityLock {
+			fromEntityLock.Lock()
+			defer fromEntityLock.Unlock()
```

Comment:
It might be a useful thing to do in many cases, but in this particular case, the lock needs to be held *almost* until the end of this function. The only thing which could get executed outside of the purview of this lock is the `txn.Commit()`, which I am thinking is nice that it is happening in the critical section. I like the idea of another write on entity storage not being successful which memDB is getting updated. Probably we should leave it as-is. Let me know.